llvm/include/llvm/Analysis/ScalarEvolution.h

```diff
@@ -1345,6 +1345,7 @@ class ScalarEvolution {
 
   class LoopGuards {
     DenseMap<const SCEV *, const SCEV *> RewriteMap;
+    DenseMap<const SCEV *, SmallPtrSet<const SCEV *, 2>> NotEqualMap;
```

Comment:
```suggestion
    DenseSet<std::pair<const SCEV *, const SCEV *>> NotEqualMap;
```
I think something like this would be simpler than the nested set?

(Could also sort the pair to avoid inserting both orders, but that's probably more hassle than it's worth.)

Reply:
Using `SmallDenseMap` here might make sense as well.